runelite-client/src/main/java/net/runelite/client/ui/ClientUI.java

```diff
@@ -280,135 +182,311 @@ public void onConfigChanged(ConfigChanged event)
+
+			titleToolbar = new ClientTitleToolbar();
+			frame.add(container);
+		});
 	}
 
-	private void init()
+	/**
+	 * Show client UI after everything else is done.
+	 *
+	 * @throws Exception exception that can occur during modification of the UI
+	 */
+	public void show() throws Exception
 	{
-		assert SwingUtilities.isEventDispatchThread();
+		final boolean withTitleBar = config.enableCustomChrome();
 
-		setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
-		addWindowListener(new WindowAdapter()
+		SwingUtilities.invokeAndWait(() ->
 		{
-			@Override
-			public void windowClosing(WindowEvent e)
+			frame.setUndecorated(withTitleBar);
+
+			if (withTitleBar)
 			{
-				checkExit();
+				frame.getRootPane().setWindowDecorationStyle(JRootPane.FRAME);
+
+				final JComponent titleBar = SubstanceCoreUtilities.getTitlePaneComponent(frame);
+				titleToolbar.putClientProperty(SubstanceTitlePaneUtilities.EXTRA_COMPONENT_KIND, SubstanceTitlePaneUtilities.ExtraComponentKind.TRAILING);
+				titleBar.add(titleToolbar);
+
+				// Substance's default layout manager for the title bar only lays out substance's components
+				// This wraps the default manager and lays out the TitleToolbar as well.
+				LayoutManager delegate = titleBar.getLayout();
+				titleBar.setLayout(new LayoutManager()
+				{
+					@Override
+					public void addLayoutComponent(String name, Component comp)
+					{
+						delegate.addLayoutComponent(name, comp);
+					}
+
+					@Override
+					public void removeLayoutComponent(Component comp)
+					{
+						delegate.removeLayoutComponent(comp);
+					}
+
+					@Override
+					public Dimension preferredLayoutSize(Container parent)
+					{
+						return delegate.preferredLayoutSize(parent);
+					}
+
+					@Override
+					public Dimension minimumLayoutSize(Container parent)
+					{
+						return delegate.minimumLayoutSize(parent);
+					}
+
+					@Override
+					public void layoutContainer(Container parent)
+					{
+						delegate.layoutContainer(parent);
+						final int width = titleToolbar.getPreferredSize().width;
+						titleToolbar.setBounds(titleBar.getWidth() - 75 - width, 0, width, titleBar.getHeight());
+					}
+				});
 			}
+
+			frame.pack();
+			SwingUtil.revalidateMinimumSize(frame);
+			frame.setLocationRelativeTo(frame.getOwner());
+			frame.setVisible(true);
+			frame.toFront();
+			requestFocus();
+			giveClientFocus();
 		});
 
-		final JPanel container = new JPanel();
-		container.setLayout(new BoxLayout(container, BoxLayout.X_AXIS));
-		container.add(new ClientPanel(client));
+		eventBus.post(new ClientUILoaded());
+	}
 
-		navContainer = new JPanel();
-		navContainer.setLayout(new BorderLayout(0, 0));
-		navContainer.setMinimumSize(new Dimension(0, 0));
-		navContainer.setMaximumSize(new Dimension(0, Integer.MAX_VALUE));
-		container.add(navContainer);
+	/**
+	 * Paint this component to target graphics
+	 *
+	 * @param graphics the graphics
+	 */
+	public void paint(final Graphics graphics)
```

Comment:
I feel it would be better to give the screenshot plugin access to the frame, rather than proxying calls for it

Reply:
That would kind of defeat the purpose of having non-swing dependant wrapper.

Reply:
I don't see the difference between handing out a Compoonent and just proxying the methods on the component. If we stop using swing it will have to be changed regardless.

Reply:
But still, exposing the internal component just for 3 methods is not very good approach, proxying it is nicer. Also, having dependency on Graphics is fine.

Reply:
Even though we don't have a hard dependency on AWT with this, it still is just AWT's API. If we go to JavaFX, we would have to change it to look like JavaFX's api. If we change to something else we would have to change to its API.

Reply:
Well the thing with the proxy is it highly restricts the usage of Swing API, to just this one AWT-specific one. It would be easier in future just to change this method than someone in some plugin in future calling methods on the internal component by misusing it.